File: runner/custom_skills.go
package runner

import (
	"fmt"
	b "mch-light/engine/battle"
	"strings"
	"unicode"
)

const CustomSkillStart uint32 = 4100000000

type CustomEffect struct {
	Kind      string `json:"kind"`
	Target    string `json:"target"`
	Chance    int32  `json:"chance"`
	Reference string `json:"reference"`
	Source    string `json:"source"`
	Basis     string `json:"basis"`
	Min       int32  `json:"min"`
	Max       int32  `json:"max"`
	Status    string `json:"status"`
	Buff      string `json:"buff"`
}
type CustomSkill struct {
	ID        uint32         `json:"id"`
	Name      string         `json:"name"`
	Trigger   string         `json:"trigger"`
	Chance    int32          `json:"chance"`
	Uses      int32          `json:"uses"`
	Threshold int32          `json:"threshold"`
	EffectID  int32          `json:"effectId"`
	Effects   []CustomEffect `json:"effects"`
}

var effectParams = map[string]b.Skill_Effect_Param{
	"damage": b.Skill_Effect_HP, "heal": b.Skill_Effect_HP, "revive": b.Skill_Effect_REVIVE,
	"phyUp": b.Skill_Effect_PHY, "phyDown": b.Skill_Effect_PHY, "intUp": b.Skill_Effect_INT, "intDown": b.Skill_Effect_INT,
	"agiUp": b.Skill_Effect_AGI, "agiDown": b.Skill_Effect_AGI, "chargeUp": b.Skill_Effect_CHARGE, "chargeDown": b.Skill_Effect_CHARGE,
	"shield": b.Skill_Effect_SHIELD, "status": b.Skill_Effect_STATUS_EFFECT, "cure": b.Skill_Effect_STATUS_EFFECT, "buff": b.Skill_Effect_BUFF_EFFECT,
}

func customTarget(key string) (*b.SkillTarget, error) {
	t := &b.SkillTarget{Team: b.Team_ALLY}
	var position b.Position
	switch key {
	case "self":
		position = b.Position_SELF
	case "previous":
		t.Team = b.Team_BOTH
		position = b.Position_TARGET
	default:
		suffix := ""
		if strings.HasPrefix(key, "enemy") {
			t.Team = b.Team_ENEMY
			suffix = strings.TrimPrefix(key, "enemy")
		} else if strings.HasPrefix(key, "ally") {
			suffix = strings.TrimPrefix(key, "ally")
		} else {
			return nil, fmt.Errorf("invalid target")
		}
		positions := map[string]b.Position{"Front": b.Position_FIRST, "Back": b.Position_LAST, "All": b.Position_ALL, "Random": b.Position_RANDOM, "Other": b.Position_EXCEPT_SELF, "SlotFront": b.Position_FRONT, "SlotMiddle": b.Position_MIDDLE, "SlotBack": b.Position_BACK}
		if p, ok := positions[suffix]; ok && !(suffix == "Other" && t.Team == b.Team_ENEMY) {
			position = p
		} else {
			params := map[string]b.Param{"LowestHp": b.Param_HP, "HighestPhy": b.Param_PHY, "HighestInt": b.Param_INT, "HighestAgi": b.Param_AGI}
			p, ok := params[suffix]
			if !ok {
				return nil, fmt.Errorf("invalid target")
			}
			calc := b.SkillTarget_ParamCondition_HIGHEST
			if suffix == "LowestHp" {
				calc = b.SkillTarget_ParamCondition_LOWEST
			}
			t.Condition = &b.SkillTarget_ParamCondition_{ParamCondition: &b.SkillTarget_ParamCondition{ParamType: b.ParamType_CURRENT, Param: p, Calc: calc}}
			return t, nil
		}
	}
	t.Condition = &b.SkillTarget_Position{Position: position}
	return t, nil
}
func CompileCustomSkill(s CustomSkill) (*b.Skill, error) {
	bad := func() (*b.Skill, error) { return nil, fmt.Errorf("invalid custom skill %d", s.ID) }
	if s.ID < CustomSkillStart || s.ID > CustomSkillStart+11 || len([]rune(strings.TrimSpace(s.Name))) < 1 || len([]rune(s.Name)) > 32 || strings.IndexFunc(s.Name, unicode.IsControl) >= 0 || len(s.Effects) < 1 || len(s.Effects) > 6 || s.Chance < 1 || s.Chance > 100 || s.Threshold < 1 || s.Threshold > 100 || s.EffectID < 1 || s.EffectID > 5 || !(s.Uses == -1 || s.Uses >= 1 && s.Uses <= 20) {
		return bad()
	}
	skill := &b.Skill{SkillId: s.ID, RemainingUses: s.Uses}
	if s.Trigger == "active" {
		if s.Uses != -1 {
			return bad()
		}
	} else {
		condition := &b.Condition{Team: b.Team_ALLY, Position: b.Position_SELF}
		if s.Trigger == "lowHp" {
			condition.Condition = &b.Condition_ParamCondition_{ParamCondition: &b.Condition_ParamCondition{Param: b.Param_HP, Calc: b.Condition_ParamCondition_UNDER, Rate: s.Threshold}}
		} else {
			states := map[string]b.State{"opening": b.State_OPENING, "damaged": b.State_AFTER_ACTIVE_SKILL_TAKEN_DAMAGE, "acted": b.State_AFTER_ACTIVE_SKILL, "abnormal": b.State_HAS_STATUS_EFFECT}
			state, ok := states[s.Trigger]
			if !ok {
				return bad()
			}
			condition.Condition = &b.Condition_StateCondition{StateCondition: state}
		}
		skill.Trigger = &b.Trigger{TriggerRate: s.Chance, Conditions: []*b.Condition{condition}}
	}
	for _, c := range s.Effects {
		param, ok := effectParams[c.Kind]
		if !ok {
			return bad()
		}
		target, err := customTarget(c.Target)
		if err != nil {
			return bad()
		}
		basis, ok := b.ParamType_value[c.Basis]
		if !ok || basis == 0 {
			return bad()
		}
		status, ok := b.StatusEffectType_value[c.Status]
		if !ok || status == 0 || c.Kind == "status" && c.Status == "ANY_EFFECT" {
			return bad()
		}
		buff, ok := b.BuffEffectType_value[c.Buff]
		if !ok || buff == 0 {
			return bad()
		}
		if c.Chance < 1 || c.Chance > 100 || c.Source != "self" && c.Source != "target" || c.Min < 0 || c.Max < c.Min || c.Max > 1000000 || c.Reference != "fixed" && c.Max > 1000 {
			return bad()
		}
		if c.Kind == "revive" && c.Target != "allyAll" && c.Target != "allySlotFront" && c.Target != "allySlotMiddle" && c.Target != "allySlotBack" {
			return bad()
		}
		e := &b.Skill_Effect{Param: param, Target: target, SuccessRate: c.Chance, IsDamage: c.Kind == "damage" || c.Kind == "status" || strings.HasSuffix(c.Kind, "Down")}
		// The sheet converter uses an exclusive maximum; this editor's displayed
		// range includes both endpoints, so convert the upper endpoint here.
		max := c.Max
		if c.Min < c.Max {
			max++
		}
		e.Value = &b.Skill_Effect_Rate_{Rate: &b.Skill_Effect_Rate{Min: c.Min, Max: max}}
		raw := &b.RawStat{Team: b.Team_ALLY, Position: b.Position_SELF}
		if c.Source == "target" {
			raw.Team = b.Team_BOTH
			raw.Position = b.Position_TARGET
		}
		switch c.Reference {
		case "fixed":
		case "HP", "PHY", "INT", "AGI":
			raw.Stat = &b.RawStat_BasicStat{BasicStat: &b.BasicStat{ParamType: b.ParamType(basis), Param: b.Param(b.Param_value[c.Reference])}}
			e.Reference = &b.Skill_Effect_RawStat{RawStat: raw}
		case "CHARGE", "DAMAGE", "HEALING", "TAKEN_DAMAGE":
			stats := map[string]b.BattleStat{"CHARGE": b.BattleStat_CHARGE, "DAMAGE": b.BattleStat_ACTION_ADDED_DAMAGE, "HEALING": b.BattleStat_ACTION_ADDED_HEALING, "TAKEN_DAMAGE": b.BattleStat_ACTIVE_SKILL_TAKEN_DAMAGE}
			raw.Stat = &b.RawStat_BattleStat{BattleStat: stats[c.Reference]}
			e.Reference = &b.Skill_Effect_RawStat{RawStat: raw}
		case "HEAL_INT":
			e.Reference = &b.Skill_Effect_ComputedStat{ComputedStat: b.ComputedStat_INT_HEALING_MODIFIER}
		case "HEAL_PHY":
			e.Reference = &b.Skill_Effect_ComputedStat{ComputedStat: b.ComputedStat_PHY_HEALING_MODIFIER}
		default:
			return bad()
		}
		if c.Kind == "status" || c.Kind == "cure" {
			e.Reference = nil
			e.Value = &b.Skill_Effect_StatusEffectType{StatusEffectType: b.StatusEffectType(status)}
		}
		if c.Kind == "buff" {
			e.Reference = nil
			e.Value = &b.Skill_Effect_BuffEffectType{BuffEffectType: b.BuffEffectType(buff)}
		}
		skill.Effects = append(skill.Effects, e)
	}
	return skill, nil
}
func customLookup(enemies []Fighter) (b.SkillLookup, error) {
	skills := map[uint32]*b.Skill{}
	for index, f := range enemies {
		if len(f.CustomSkills) > 4 {
			return nil, fmt.Errorf("too many custom skills")
		}
		owned := map[uint32]*b.Skill{}
		for _, c := range f.CustomSkills {
			if c.ID < CustomSkillStart+uint32(index*4) || c.ID > CustomSkillStart+uint32(index*4+3) {
				return nil, fmt.Errorf("custom skill belongs to another fighter")
			}
			s, err := CompileCustomSkill(c)
			if err != nil {
				return nil, err
			}
			if skills[c.ID] != nil {
				return nil, fmt.Errorf("duplicate custom skill")
			}
			skills[c.ID] = s
			owned[c.ID] = s
		}
		for _, id := range f.Actives {
			if id >= CustomSkillStart {
				s := owned[id]
				if s == nil || s.Trigger != nil {
					return nil, fmt.Errorf("invalid custom active")
				}
			}
		}
		if f.Passive >= CustomSkillStart {
			s := owned[f.Passive]
			if s == nil || s.Trigger == nil {
				return nil, fmt.Errorf("invalid custom passive")
			}
		}
	}
	return func(id uint32) (*b.Skill, error) {
		if s := skills[id]; s != nil {
			return s, nil
		}
		return b.GetSkill(id)
	}, nil
}

File: shared/custom-skills.ts
import {
  compileCustomEffect,
  compileCustomSkill,
} from "./custom-skill-mechanics";
import { describeBattleEffect, describeBattleSkill } from "./skill-description";
// The editor labels and replay descriptions share one vocabulary. These keys
// are a bounded data format, never executable code or user supplied formulas.
export const CUSTOM_SKILL_START = 4100000000;
export const kinds = {
  damage: "ダメージ",
  heal: "HPを回復",
  revive: "倒れた味方を復活",
  phyUp: "PHYを上げる",
  phyDown: "PHYを下げる",
  intUp: "INTを上げる",
  intDown: "INTを下げる",
  agiUp: "AGIを上げる",
  agiDown: "AGIを下げる",
  chargeUp: "チャージを増やす",
  chargeDown: "チャージを減らす",
  shield: "シールドを張る",
  status: "状態を付与",
  cure: "状態異常を解除",
  buff: "特殊効果を付与",
} as const;
export const targets = {
  self: "自分",
  enemyFront: "敵の先頭",
  enemyBack: "敵の最後尾",
  enemyAll: "敵全体",
  enemyRandom: "ランダムな敵1体",
  enemyLowestHp: "残りHPが最も低い敵",
  enemyHighestPhy: "PHYが最も高い敵",
  enemyHighestInt: "INTが最も高い敵",
  enemyHighestAgi: "AGIが最も高い敵",
  allyFront: "味方の先頭",
  allyBack: "味方の最後尾",
  allyAll: "味方全体",
  allyOther: "自分以外の味方",
  allyRandom: "ランダムな味方1体",
  allyLowestHp: "残りHPが最も低い味方",
  allyHighestPhy: "PHYが最も高い味方",
  allyHighestInt: "INTが最も高い味方",
  allyHighestAgi: "AGIが最も高い味方",
  allySlotFront: "味方の前衛（位置指定）",
  allySlotMiddle: "味方の中衛（位置指定）",
  allySlotBack: "味方の後衛（位置指定）",
  previous: "このスキルでダメージを与えた相手",
} as const;
export const references = {
  fixed: "固定の数値",
  PHY: "PHY",
  INT: "INT",
  AGI: "AGI",
  HP: "HP",
  CHARGE: "チャージ",
  DAMAGE: "この行動で与えたダメージ",
  HEALING: "この行動で回復したHP",
  TAKEN_DAMAGE: "直前の行動で受けたダメージ",
  HEAL_INT: "回復係数（自身のINTと対象のPHYの平均）",
  HEAL_PHY: "PHY回復係数（自身のPHYと対象のINTの平均）",
} as const;
export const bases = {
  CURRENT: "現在の値",
  BASE: "戦闘開始時の値",
  INCREASE: "増えた分",
  DECREASE: "減った分",
} as const;
export const statuses = {
  POISON: "毒",
  SLEEP: "睡眠",
  CONFUSION: "混乱",
  FEAR: "恐怖",
  BARRIER: "バリア",
  BLEED: "出血",
  STUN: "気絶",
  PROSTRATION: "衰弱",
  CURSE: "呪い",
  CHARM: "魅了",
  BIND: "バインド",
  GRAVITY: "重力",
} as const;
export const buffs = {
  RESURRECTION: "復活予約",
  DECOY: "デコイ（攻撃を引き受ける）",
  AGI_DAMAGE_REDUCTION: "AGIを使ったダメージの軽減",
  HP_DAMAGE_REDUCTION: "HPを使ったダメージの軽減",
} as const;
export const triggers = {
  active: "行動順で使う",
  opening: "バトル開始時",
  damaged: "行動スキルでダメージを受けた後",
  acted: "自分の行動後",
  lowHp: "自分のHPが指定割合を下回った時",
  abnormal: "自分が状態異常の時",
} as const;
export const visuals = {
  1: "物理攻撃",
  2: "魔法攻撃",
  3: "回復",
  4: "強化",
  5: "弱体",
} as const;
export type CustomEffect = {
  kind: keyof typeof kinds;
  target: keyof typeof targets;
  chance: number;
  reference: keyof typeof references;
  source: "self" | "target";
  basis: keyof typeof bases;
  min: number;
  max: number;
  status: keyof typeof statuses | "ANY_EFFECT";
  buff: keyof typeof buffs;
};
export type CustomSkill = {
  id: number;
  name: string;
  trigger: keyof typeof triggers;
  chance: number;
  uses: number;
  threshold: number;
  effectId: number;
  effects: CustomEffect[];
};
export const defaultEffect = (): CustomEffect => ({
  kind: "damage",
  target: "enemyFront",
  chance: 100,
  reference: "PHY",
  source: "self",
  basis: "CURRENT",
  min: 30,
  max: 35,
  status: "POISON",
  buff: "DECOY",
});
export const defaultSkill = (id: number, passive = false): CustomSkill => ({
  id,
  name: "新しいスキル",
  trigger: passive ? "opening" : "active",
  chance: 100,
  uses: passive ? 1 : -1,
  threshold: 50,
  effectId: 1,
  effects: [defaultEffect()],
});
const own = (values: object, v: unknown) =>
  typeof v === "string" && Object.hasOwn(values, v);
const int = (v: unknown, min: number, max: number) =>
  Number.isInteger(v) && Number(v) >= min && Number(v) <= max;
export function parseCustomSkill(value: any): CustomSkill {
  const s = value;
  if (
    !s ||
    !int(s.id, CUSTOM_SKILL_START, CUSTOM_SKILL_START + 11) ||
    typeof s.name !== "string" ||
    !s.name.trim() ||
    s.name.trim().length > 32 ||
    /[\u0000-\u001f\u007f]/.test(s.name)
  )
    throw Error("スキルの名前は1〜32文字で入力してください");
  if (
    !own(triggers, s.trigger) ||
    !int(s.chance, 1, 100) ||
    !(s.uses === -1 || int(s.uses, 1, 20)) ||
    !int(s.threshold, 1, 100) ||
    !int(s.effectId, 1, 5)
  )
    throw Error("発動条件・確率・回数を確認してください");
  if (s.trigger === "active" && s.uses !== -1)
    throw Error("行動スキルは回数制限なしで設定してください");
  if (!Array.isArray(s.effects) || s.effects.length < 1 || s.effects.length > 6)
    throw Error("効果は1〜6個で作ってください");
  const effects = s.effects.map((e: any): CustomEffect => {
    if (
      !e ||
      !own(kinds, e.kind) ||
      !own(targets, e.target) ||
      !int(e.chance, 1, 100) ||
      !own(references, e.reference) ||
      !["self", "target"].includes(e.source) ||
      !own(bases, e.basis)
    )
      throw Error("効果の対象・内容・確率を確認してください");
    if (
      (!own(statuses, e.status) && e.status !== "ANY_EFFECT") ||
      !own(buffs, e.buff)
    )
      throw Error("状態変化を選んでください");
    if (e.kind === "status" && e.status === "ANY_EFFECT")
      throw Error("付与する状態を選んでください");
    if (
      !int(e.min, 0, 1000000) ||
      !int(e.max, e.min, 1000000) ||
      (e.reference !== "fixed" && e.max > 1000)
    )
      throw Error("数値は0〜1,000,000、倍率は0〜1,000%で設定してください");
    if (
      e.kind === "revive" &&
      !["allyAll", "allySlotFront", "allySlotMiddle", "allySlotBack"].includes(
        e.target,
      )
    )
      throw Error("復活の対象は自分・味方から選んでください");
    return {
      kind: e.kind,
      target: e.target,
      chance: e.chance,
      reference: e.reference,
      source: e.source,
      basis: e.basis,
      min: e.min,
      max: e.max,
      status: e.status,
      buff: e.buff,
    };
  });
  return {
    id: s.id,
    name: s.name.trim(),
    trigger: s.trigger,
    chance: s.chance,
    uses: s.uses,
    threshold: s.threshold,
    effectId: s.effectId,
    effects,
  };
}
export function describeEffect(e: CustomEffect): string {
  return describeBattleEffect(compileCustomEffect(e), { custom: true });
}
export function customSkillDetail(s: CustomSkill) {
  const raw = compileCustomSkill(s),
    ja = describeBattleSkill(raw, s, { custom: true }),
    en = describeBattleSkill(raw, s, { custom: true, language: "en" });
  return {
    ...ja,
    en: { name: en.name, condition: en.condition, effects: en.effects },
  };
}
export const customSkillCatalog = (team: { customSkills?: CustomSkill[] }[]) =>
  team.flatMap((f) => (f.customSkills || []).map(customSkillDetail));
